src/main/java/fi/metatavu/metamind/server/keycloak/AuthenticationController.java

```diff
@@ -0,0 +1,373 @@
+    
+    if (foundResourcesIds.isEmpty()) {
+      return null;
+    }
+    
+    return UUID.fromString(foundResourcesIds.get(0));
+  }
+
+  /**
+   * Creates new scope permission for resource
+   * 
+   * @param resourceId UUID resourceId
+   * @param scopes AuthorizationScope collection
+   * @param name String name
+   * @param decisionStrategy decisionStrategy
+   * @param policyIds UUID collection
+   * @throws AuthorizationException
+   * @return String permission name
+   */
+  public String upsertScopePermission(UUID resourceId, Collection<AuthorizationScope> scopes, String name, DecisionStrategy decisionStrategy,
+      Collection<UUID> policyIds) throws AuthorizationException {
+    Keycloak keycloak = getAdminClient();
+    ClientRepresentation client = getClient(keycloak);
+    String realmName = getRealmName();
+    
+    if (client == null) {
+      throw new AuthorizationException(AUTHORIZATION_EXCEPTION_MESSAGE);
+    }
+    
+    RealmResource realm = keycloak.realm(realmName);
+    ScopePermissionsResource scopeResource = realm.clients().get(client.getId()).authorization().permissions().scope();
+    ScopePermissionRepresentation existingPermission = scopeResource.findByName(name);
+
+    ScopePermissionRepresentation representation = new ScopePermissionRepresentation();
+    representation.setResources(Collections.singleton(resourceId.toString()));
+    representation.setDecisionStrategy(decisionStrategy);
+    representation.setName(name);
+    representation.setScopes(scopes.stream().map(AuthorizationScope::getName).collect(Collectors.toSet()));
+    representation.setPolicies(policyIds.stream().map(UUID::toString).collect(Collectors.toSet()));
+
+    String result = representation.getName();
+    Response response = scopeResource.create(representation);
+    try {
+      
+      if (existingPermission == null) {
+        int status = response.getStatus();
+        
+        if (status != 201) {
+          String message = "Unknown error";
+          
+          try {
+            message = IOUtils.toString((InputStream) response.getEntity(), StandardCharsets.UTF_8);
+          } catch (IOException e) {
+            logger.warn("Failed read error message", e);
+          }
+
+          logger.warn("Failed to create scope permission for resource {} with message {}", resourceId, message);
+        }
+        
+      } else {
+        realm.clients().get(client.getId()).authorization().permissions().scope().findById(existingPermission.getId()).update(representation);
+      }
+      
+    } finally {
+      response.close();
+    }
+    
+    return result;
+  }
+
+  /**
+   * Updates users and user policies into Keycloak
+   * 
+   * @param keycloak admin client
+   * @param realmName realm name
+   * @param clientId client id
+   * @param userMap users names
+   */
+  public List<UUID> checkUserPolicy(UUID loggedUserId) throws AuthorizationException {
```

Comment:
checkUserPolicy name is a bit confusing, maybe use something like ensureUserPolicyExists